contrib/nvim_api_docs.py

```diff
@@ -0,0 +1,527 @@
+    p.communicate(config.format(input=src_dir, output=out_dir,
+                                filter=filter_cmd).encode('utf8'))
+    if p.returncode:
+        sys.exit(p.returncode)
+
+    title_length = 0
+    toc = []
+    sections = []
+    sep = '=' * text_width
+
+    base = os.path.join(out_dir, 'xml')
+    dom = minidom.parse(os.path.join(base, 'index.xml'))
+    for compound in dom.getElementsByTagName('compound'):
+        if compound.getAttribute('kind') != 'file':
+            continue
+
+        filename = get_text(find_first(compound, 'name'))
+        if filename.endswith('.c'):
+            functions, deprecated = parse_source_xml(
+                os.path.join(base, '%s.xml' % compound.getAttribute('refid')))
+
+            if functions or deprecated:
+                name = os.path.splitext(os.path.basename(filename))[0]
+                if name == 'ui':
+                    name = name.upper()
+                else:
+                    name = name.title()
+
+                title = '%d. %s Functions' % (len(toc) + 1, name)
+                helptag = '*nvim-api-%s*' % name.lower()
+                title_length = max(title_length, len(title))
+                toc.append((title, helptag))
+
+                doc = sep
+                doc += '\n' + title + helptag.rjust(text_width - len(title))
+                if functions:
+                    doc += '\n\n' + functions
+
+                if 'INCLUDE_DEPRECATED' in os.environ and deprecated:
+                    doc += '\n\n\nDeprecated %s Functions:~\n\n' % name
+                    doc += deprecated
+
+                sections.append(doc)
+
+    if not toc:
+        return
+
+    title_left = '*api-funcs.txt*'
+    title_center = 'Neovim API Function Reference'
+    title_right = '{Nvim}'
+    margin = max(len(title_left), len(title_right))
+    doc = (title_left.ljust(margin) +
+           title_center.center(text_width - margin * 2) +
+           title_right.rjust(margin)) + '\n'
+
+    doc += '\n%s\n\n' % doc_wrap(preamble, width=text_width)
+    doc += 'Contents:\n\n'
+
+    for title, helptag in toc:
+        doc += title.ljust(title_length) + '  ' + helptag.replace('*', '|') + '\n'
+
+    doc += '\n'
+    for section in sections:
+        doc += section
+        doc += '\n\n\n'
+
+    doc = doc.rstrip() + '\n\n'
+    doc += ' vim:tw=78:ts=8:ft=help:norl:'
+
+    doc_file = os.path.join(base_dir, 'runtime/doc/api-functions.txt')
+    with open(doc_file, 'wb') as fp:
+        fp.write(doc.encode('utf8'))
+
+    shutil.rmtree(out_dir)
+
+
+def filter_source(filename):
+    """Filters the source to fix macros that confuse Doxygen."""
+    with open(filename, 'rt') as fp:
+        print(re.sub(r'^(ArrayOf|DictionaryOf)(\(.*?\))',
+                     lambda m: m.group(1)+'_'.join(
+                         re.split(r'[^\w]+', m.group(2))),
+                     fp.read(), flags=re.M))
+
+
+# Doxygen Config {{{
```

Comment:
Instead of duplicating, it wouldn't be possible to instead change the global Doxyfile in `src/` ? The filtering would also be useful for the html output, but it would then need to instead be done in `lua` or `sed` (or some other build dependency).